test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -212,34 +218,257 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+
+		ginkgo.By("deleting the snapshot")
+		// typically this snapshot has already been deleted
+		err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Delete(context.TODO(), snapshot.GetName(), metav1.DeleteOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			framework.Failf("Error deleting snapshot %s in namespace %s. Error: %v", snapshot.GetName(), snapshot.GetNamespace(), err)
+		}
+
+		ginkgo.By("checking the SnapshotContent has been deleted")
+		err = WaitForSnapshotContentDelete(dc, snapshotContentName, framework.Poll, framework.SnapshotDeleteTimeout)
+		framework.ExpectNoError(err)
+	})
+
+	ginkgo.It("should not delete snapshot with retain policy [Feature:VolumeSnapshotDataSource]", func() {
+		cs := f.ClientSet
+		dc := f.DynamicClient
+
+		// Now do the more expensive test initialization.
+		config, driverCleanup := driver.PrepareTest(f)
+		defer func() {
+			err := tryFunc(driverCleanup)
+			framework.ExpectNoError(err, "while cleaning up driver")
+		}()
+
+		vsc := sDriver.GetSnapshotClass(config)
+		class := dDriver.GetDynamicProvisionStorageClass(config, "")
+		if class == nil {
+			e2eskipper.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", driver.GetDriverInfo().Name)
+		}
+		testVolumeSizeRange := s.GetTestSuiteInfo().SupportedSizeRange
+		driverVolumeSizeRange := dDriver.GetDriverInfo().SupportedSizeRange
+		claimSize, err := getSizeRangesIntersection(testVolumeSizeRange, driverVolumeSizeRange)
+		framework.ExpectNoError(err, "determine intersection of test size range %+v and driver size range %+v", testVolumeSizeRange, driverVolumeSizeRange)
+		pvc := e2epv.MakePersistentVolumeClaim(e2epv.PersistentVolumeClaimConfig{
+			ClaimSize:        claimSize,
+			StorageClassName: &(class.Name),
+		}, config.Framework.Namespace.Name)
+
+		framework.Logf("In creating storage class object and pvc object for driver - sc: %v, pvc: %v", class, pvc)
+
+		ginkgo.By("creating a StorageClass " + class.Name)
+		class, err = cs.StorageV1().StorageClasses().Create(context.TODO(), class, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		defer func() {
+			framework.Logf("deleting storage class %s", class.Name)
+			framework.ExpectNoError(cs.StorageV1().StorageClasses().Delete(context.TODO(), class.Name, metav1.DeleteOptions{}))
+		}()
+
+		ginkgo.By("creating a claim")
+		pvc, err = cs.CoreV1().PersistentVolumeClaims(pvc.Namespace).Create(context.TODO(), pvc, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		defer func() {
+			framework.Logf("deleting claim %q/%q", pvc.Namespace, pvc.Name)
+			// typically this claim has already been deleted
+			err = cs.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(context.TODO(), pvc.Name, metav1.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				framework.Failf("Error deleting claim %q. Error: %v", pvc.Name, err)
+			}
+		}()
+
+		ginkgo.By("starting a pod to use the claim")
```

Comment:
All of these steps seem similar to other test cases and can probably benefit from some helper utilities to reduce code duplication and make the test case easier to read.

Reply:
Makes sense, will clean up these tests with helper utilities.